hs/src/Reach/Eval/Core.hs

```diff
@@ -1088,22 +1089,25 @@ infectWithId_clo :: SLVar -> SLClo -> SLClo
 infectWithId_clo v (SLClo _ e b c) =
   SLClo (Just v) e b c
 
-infectWithId_sv :: SrcLoc -> SLVar -> SLVal -> SLVal
-infectWithId_sv at v = \case
-  SLV_Participant a who _ mdv ->
-    SLV_Participant a who (Just v) mdv
-  SLV_Clo a mt c ->
-    SLV_Clo a mt $ infectWithId_clo v c
-  SLV_DLVar (DLVar a _ t i) ->
-    SLV_DLVar $ DLVar a (Just (at, v)) t i
-  x -> x
-
-infectWithId_sss :: SLVar -> SLSSVal -> SLSSVal
-infectWithId_sss v (SLSSVal at lvl sv) =
-  SLSSVal at lvl $ infectWithId_sv at v sv
-
-infectWithId_sls :: SrcLoc -> SLVar -> SLSVal -> SLSVal
-infectWithId_sls at v (lvl, sv) = (lvl, infectWithId_sv at v sv)
+infectWithId_sv :: SrcLoc -> SLVar -> SLVal -> App SLVal
+infectWithId_sv at v val = do
+  case val of
+    SLV_Participant a who _ mdv -> return $ SLV_Participant a who (Just v) mdv
+    SLV_Clo a mt c -> return $ SLV_Clo a mt $ infectWithId_clo v c
+    SLV_DLVar (DLVar a _ t i) -> do
+      infections <- asks e_infections
+      liftIO $ modifyIORef infections $ M.insert (fromIntegral i) (at, v)
```

Comment:
You insert and extract with a `fromIntegral`, so you should just change the type